### PR TITLE
DevWs for Most Target Limited containing containing Zigbee Multi Button

### PR DESCRIPTION
**Pull request by adurosmart**

### CODE REVIEW

**Comment by adurosmart** [comment]
@greens Help me to review this code,we add code of dimming button and scene button ,Thank you very much.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: you'll need to add the 

`, mnmn: "SmartThings", vid: "generic-4-button"`

from the above lines to have this display correctly.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: it would be better to add this inside `parseAttrMessage` to maintain the parallel structure.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: you've used this same conditional in several places. It's probably best for readability to break it out into its own function. Perhaps `isAduroSmartRemote()`?

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Please use constants instead of magic numbers for clusters.

0x0006 is `zigbee.ONOFF_CLUSTER`
0x0008 is `zigbee.LEVEL_CONTROL_CLUSTER`

for 0xFCCC define a function 

`def getMANUFACTURER_SPECIFIC_CLUSTER() {0xFCCC}`

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:254]: I'm worried that since your devices report using the level control cluster and you're not setting up any bindings here that they may not actually report when they are pressed.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: this value will never be anything except "pushed"

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: `buttonNumber = (list2[1] as int) + 1`

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: these lines here and above are only used when the device sends a different message when the button is pressed down and when it is released so that we can emulate "pushed" and "held". You are only doing "pushed" here, so this is unnecessary.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: you duplicate these exact 3 lines of code in 3 separate if statements instead of just putting it after them

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:254]: same for 0xFCCC

**Review by greens — CHANGES_REQUESTED** [review]
On top of specific requests, please use tabs instead of spaces to match the existing file.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: using spaces instead of tabs

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: thanks to some nice syntactic sugar in groovy, you can just type `AduroSmart_SPECIFIC_CLUSTER` without the `get` or `()`. It's essentially the Groovy way to do constants, which is why the whole function name (except for `get`) is typically capitalized.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: just so you know, your display names will be absurdly long and most likely be truncated, making it difficult to determine which button is which. For example:

`Eria scene button switch V2.0 child button 1`

Perhaps consider changing your join names to `Eria Remote` or something like that. If you like it how it is, that's fine, I'm just warning you about usability concerns I have.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: see: `getButtonMap()`

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Reconsider the placement of this `if` statement. I'm worried it might result in errors when the device reports its battery level.

Comment (greens) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: And the results will be overwritten when the device sends ONOFF messages.

It might fit as an `else if` after the power config cluster statement

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: the code using `getButtonMap` is not intended to be used in the flow you've outlined I'm pretty sure, making these additions unnecessary.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: @greens 
Thanks.
Fixed.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: i using " - " instead of "child button"

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: fixed.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Fixed.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: nit: spaces instead of tabs again

your editor should have a function that allows you to convert spaces->tabs all at once

**Comment by greens** [comment]
@tpmanley before I sign off on this, can you take a look?

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: nit: please put spaces after the `}`

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: This line isn't needed because it's already done at the top of the function

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: nit: this could be an `else if` instead

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: nit: should have a space after the `!=`

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: nit: should have a space before `data`

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: nit: spaces before `else if` in this function too

**Review by tpmanley — DISMISSED** [review]
Just a few nits but functionally it looks good

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Fixed issues

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Fixed issues

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Fixed issues

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Fixed issues

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Fixed issues

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Fixed issues

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (tpmanley) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: @adurosmart the line `def descMap = zigbee.parseDescriptionAsMap(description)` is still present but should be removed

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (tpmanley) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: @adurosmart this is still an `if` but should be changed to an `else if`

Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Thank you review,Fixed issues.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: @tpmanley Thank you review,Fixed issues.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (tpmanley) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:254]: @adurosmart did you take a look at this comment? I think you at least need the level control binding and likely the 0xFCCC binding unless you handle that cluster automatically without a binding.

Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:254]: @tpmanley  Thank you review, i add LEVEL_CONTROL_CLUSTER  bindings.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Do you mean to reference `endpoint` here? `index` is never incremented

Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: @tpmanley, Thank you review, i changed the index increment method.

Comment (tpmanley) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: @adurosmart I think it would be better to avoid the introduction of an additional `index` variable and just use `endpoint`.  Alternatively, you could leave the `for` loop untouched and put the following after the `for` loop:

```
if (isAduroSmartRemote()) {
    bindings += zigbee.addBinding(zigbee.LEVEL_CONTROL_CLUSTER, ["destEndpoint" : 2]) + 
        zigbee.addBinding(zigbee.LEVEL_CONTROL_CLUSTER, ["destEndpoint" : 3])
}
```

**Comment by shrakuma** [comment]
@adurosmart  can you please take a look at the above

**Comment by tpmanley** [comment]
@shrakuma they were waiting on us to review the latest changes. I just missed that updates had been made.

@adurosmart the changes look good to me now. Unfortunately there is now a merge conflict due to other changes to the same file. Can you rebase your changes on the latest version and then ping me to do a final review and merge?

**Comment by adurosmart** [comment]
@tpmanley  
Thank you review.
Is it using the 'git rebase master' command?
Thank you very much for your suggestions.

**Comment by tpmanley** [comment]
@adurosmart yes it can be done with `rebase`, but you might find it easier to just manually make the changes to the latest version of the file. To do that, first make a copy of the existing file, then do this:

```
git fetch -p
git reset --hard upstream/master # This will remove your local changes so make sure you have a copy saved
<manually update the file with the changes from your saved copy>
<commit changes>
git push -f origin devws-for-most-target-limited-co_144
```

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: this code has been duplicated

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:294]: nit: spacing

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: nit: spacing

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (greens) [inline devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: I don't think you need to short-circuit this function call here, you should be handling this correctly below.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Fixed

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy:294]: Thank you review.
In these few lines of code, I didn't find any problems with word spacing. Maybe I didn't see it.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Thank you review.
Fixed.

File: devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy
Comment (adurosmart) [reply devicetypes/smartthings/zigbee-multi-button.src/zigbee-multi-button.groovy]: Thank you review.
Fixed.

**Review by greens — APPROVED** [review]

**Comment by Automated-House** [comment]
Hi @adurosmart I've paired a Eria dimming button switch V2.0 to SmartThings using this now published DTH. There's a couple problems:
1) Device isn't checking in to the hub at the `checkInterval` defined in the DTH, so it's being marked as offline in the SmartThings app
2) Battery isn't reporting correctly.

**Pull request merged**
